File: STL_VS_PY/dict.cpp
#include <iostream>
#include <unordered_map>
#include <chrono>

int main() {
    const int N = 1000000;
    std::unordered_map<int, int> d;

    // Insert
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < N; ++i) {
        d[i] = i;
    }
    auto end = std::chrono::high_resolution_clock::now();
    std::cout << "C++ unordered_map insert time: "
              << std::chrono::duration<double>(end - start).count() << "s\n";

    // Lookup
    start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < N; ++i) {
        int x = d[i];
    }
    end = std::chrono::high_resolution_clock::now();
    std::cout << "C++ unordered_map lookup time: "
              << std::chrono::duration<double>(end - start).count() << "s\n";
}

File: STL_VS_PY/dict.py
import time

def insert_benchmark(d, N):
    for i in range(N):
        d[i] = i

def lookup_benchmark(d, N):
    for i in range(N):
        _ = d[i]

def main():
    N = 1_000_000
    d = {}

    start = time.time()
    insert_benchmark(d, N)
    mid = time.time()
    lookup_benchmark(d, N)
    end = time.time()

    print(f"Python dict insert time: {mid - start}")
    print(f"Python dict lookup time: {end - mid}")

if __name__ == "__main__":
    main()
